### PR TITLE
Plugins: Fix Default Nav URL for dashboard includes

### LINKED ISSUE #46412: Default Nav links do not work for dashboards in App Plugins in Grafana 8+
**What is happening?**
App plugins having defaultNav value configured for one of the included dashboards have a broken link in the sidebar, when clicking on the App Icon. See GIF below on how it currently behaves.

[Gif/Video of how it behaves](https://i.imgur.com/l6uH8AY.gifv)

**Environments affected**
Grafana `8.+`

**How to reproduce**
Create an app plugin and have one of the dashboard configured as a `defaultNav`. 

```
  "includes": [
    {
      "type": "dashboard",
      "name": "Zones",
      "path": "dashboards/dashboard_example.json",
      "uid": "KDVdMCw2k",
      "addToNav": true,
      "defaultNav": true
    },
```

**The reason it happens**
The local `/api/plugins?embedded=0&core=0`

returns the wrong `defaultNavUrl:` 

```
[
    {
        "name": "Grafana App",
        "type": "app",
        "id": "grafana-app",
        "enabled": true,
        "pinned": true,
        "info": {
            "author": {
                "name": "Grafana",
                "url": "https://www.grafana.com/"
            },
            "description": "Grafana App ",
            "links": [
                {
                    "name": "Project site",
                    "url": "https://github.com/grafana"
                },
                {
                    "name": "License \u0026 Terms",
                    "url": "https://github.com/grafana/blob/master/LICENSE"
                }
            ],
            "logos": {
                "small": "public/plugins/grafana-app/img/grafana_icon.png",
                "large": "public/plugins/grafana-app/img/grafana_icon.png"
            },
            "build": {},
            "screenshots": [
                {
                    "name": "Dashboard: 1",
                    "path": "public/plugins/grafana-app/img/grafana_dashboard_1.png"
                },
                {
                    "name": "Dashboard: 2",
                    "path": "public/plugins/grafana-app/img/grafana_dashboard_2.png"
                }
            ],
            "version": "0.0.1",
            "updated": "2022-03-10"
        },
        "dependencies": {
            "grafanaDependency": ">=8",
            "grafanaVersion": "8.x.x",
            "plugins": []
        },
        "latestVersion": "",
        "hasUpdate": false,
        "defaultNavUrl": "/dashboard/db/1",
        "category": "",
        "state": "",
        "signature": "unsigned",
        "signatureType": "",
        "signatureOrg": ""
    }
]
```

The backend code responsible for that part is:
```
if include.Type == "dashboard" {
	p.DefaultNavURL = path.Join("/dashboard/db/", include.Slug)
}
```

### PR DESCRIPTION
**Pull request by wbrowne**
**Which issue(s) this PR fixes**:

Fixes #46412

### CODE REVIEW

File: pkg/plugins/manager/loader/loader.go [hunk 1]
```diff
@@ -271,7 +271,12 @@ func setDefaultNavURL(p *plugins.Plugin) {
 			p.DefaultNavURL = path.Join("/plugins/", p.ID, "/page/", include.Slug)
 		}
 		if include.Type == "dashboard" {
-			p.DefaultNavURL = path.Join("/dashboard/db/", include.Slug)
+			if include.UID == "" {
+				p.Logger().Warn("Included dashboard is missing a UID field")
```
Comment (tolzhabayev) [inline pkg/plugins/manager/loader/loader.go:275]: Falling back to `path.Join("/dashboard/db/", include.Slug)` makes no sense, right?

Comment (wbrowne) [reply pkg/plugins/manager/loader/loader.go:275]: Sadly not. That API is long done :rip:

File: pkg/plugins/manager/loader/loader_test.go [hunk 2]
```diff
@@ -378,7 +378,7 @@ func TestLoader_Load(t *testing.T) {
 						Plugins:           []plugins.Dependency{},
 					},
 					Includes: []*plugins.Includes{
-						{Name: "Nginx Memory", Path: "dashboards/memory.json", Type: "dashboard", Role: "Viewer", Slug: "nginx-memory", DefaultNav: true},
+						{Name: "Nginx Memory", Path: "dashboards/memory.json", Type: "dashboard", Role: "Viewer", Slug: "nginx-memory"},
```
Comment (wbrowne) [inline pkg/plugins/manager/loader/loader_test.go:381]: I updated this purely as this is technically invalid, as both includes have `DefaultNav: true` - we should add validation earlier in the process to formally prevent this (plugin publishing, and also during the loading step)

**Review by tolzhabayev — APPROVED** [review]
LGTM! 🍔

File: pkg/plugins/manager/loader/loader.go (hunk 1 above)
Comment (marefr) [reply pkg/plugins/manager/loader/loader.go:275]: There's some related code here which does fallback to slug for whatever reason: https://github.com/grafana/grafana/blob/1a4b1184bd6fcf73d5749edb571d0e628f654078/pkg/api/index.go#L122-L129

**Review by marefr — COMMENTED** [review]
Added a comment

File: pkg/api/index.go [hunk 3]
```diff
@@ -120,11 +120,15 @@ func (hs *HTTPServer) getAppLinks(c *models.ReqContext) ([]*dtos.NavLink, error)
 			}
 
 			if include.Type == "dashboard" && include.AddToNav {
-				link := &dtos.NavLink{
-					Url:  hs.Cfg.AppSubURL + include.GetSlugOrUIDLink(),
-					Text: include.Name,
+				dboardPath := include.DashboardURLPath()
+
+				if dboardPath != "" {
```
Comment (marefr) [inline pkg/api/index.go:125]: Maybe a warning logged here as well if it's empty?

File: pkg/plugins/manager/loader/loader.go [hunk 4]
```diff
@@ -271,7 +271,12 @@ func setDefaultNavURL(p *plugins.Plugin) {
 			p.DefaultNavURL = path.Join("/plugins/", p.ID, "/page/", include.Slug)
 		}
 		if include.Type == "dashboard" {
-			p.DefaultNavURL = path.Join("/dashboard/db/", include.Slug)
+			if include.UID == "" {
```
Comment (marefr) [inline pkg/plugins/manager/loader/loader.go:274]: Want to check if `include.DashboardURLPath()` is empty like in index.go? Then maybe we need to change the message?

**Review by marefr — COMMENTED** [review]
Looks good, but added some additional comments

File: pkg/api/index.go (hunk 3 above)
Comment (wbrowne) [reply pkg/api/index.go]: I think this will be too spammy, as this runs on every index refresh

Comment (marefr) [reply pkg/api/index.go]: Good point :+1:

**Review by marefr — APPROVED** [review]
LGTM

**Pull request merged**